File: rust/src/bin/p0096.rs
// Project Euler: Problem 96

/*
  This solution uses a simple backtracking algorithm.

      C0 C1 C2 C3 C4 C5 C6 C7 C8
     +--------+--------+--------+    R: Row
  R0 | 0  1  2| 3  4  5| 6  7  8|    C: Column
  R1 | 9 10 11|12      |15      |
  R2 |18      |21      |24      |    Cell: 0..=80
     +--------+--------+--------+
  R3 |27      |30      |33      |
  R4 |36      |39      |42      |
  R5 |45      |48      |51      |
     +--------+--------+--------+
  R6 |54      |57      |60      |
  R7 |63      |66      |69      |
  R8 |72      |75 76 77|78 79 80|
     +--------+--------+--------+

  grid: [u128; 10]
    grid[n] is the information of number 'n'

    127  81 80                   0
    |......|.....................|
     unused       bit table

  Example:
   * Grid 01 (from p096_sudoku.txt)
      003|020|600
      900|305|001
      001|806|400
      ---+---+---
      008|102|900
      700|000|008
      006|708|200
      ---+---+---
      002|609|500
      800|203|009
      005|010|300

    Initial Data: grid[]
      [0x1ab6b64f26fec9e4dadab,
       0x010000000000040120000, 0x000040108000100000010, 0x040100000000000001004,
       0x000000000000001000000, 0x004001000000000004000, 0x000000200800000800040,
       0x000000001001000000000, 0x000008004100020200000, 0x000800800000200000200]

    1) number '0'
        python> bin(grid[0])[::-1][0:10]    # print only 'R0'
        '110101011'

        R0: 003|020|600
           '110 101 011'

    2) number '3'
        python> bin(grid[3])[::-1][0:10]    # print only 'R0'
        '001000000'

        R0: 003|020|600
           '001 000 000'

   * Grid 01 (Solution)
      483|921|657
      967|345|821
      251|876|493
      ---+---+---
      548|132|976
      729|564|138
      136|798|245
      ---+---+---
      372|689|514
      814|253|769
      695|417|382

    Solution: sol[]
      [0x1ab6b64f26fec9e4dadab,
       0x010012000240040120020, 0x100040108002100050010, 0x040100040480084001004,
       0x008024010020011002001, 0x004081020008008084080, 0x001400200810800800440,
       0x020200081001400400900, 0x080008404100020208002, 0x002800802004202000208]

    1) number '4'
      python> "{:081b}".format(sol[4])[::-1]
     '100000000000010000000000100010000000000001000000000010000000001001000000000100000'
      python> s = "{:081b}".format(sol[4])[::-1]
      python> for i in range(0, len(s), 9):
                  print(s[i:i+9])

      100000000
      000010000
      000000100
      010000000
      000001000
      000000010
      000000001
      001000000
      000100000

        The set bits indicate that there is a number at position. In this case, the number is 4.
        Please compare it with the solution of Grid 01.
*/

use regex::Regex;

euler::run_solver!(96);

static FILE_DATA: &str = include_str!("../../assets/0096_sudoku.txt");

const NUM_OF_CELLS: usize = 81;

// 0b: 111111111
const ROW_MASK: u128 = 0x1ff;
// 0b: 000000001 000000001 ... 000000001
const COL_MASK: u128 = 0x1008040201008040201;
// 0b: 000000111 000000111 000000111
const BOX_MASK: u128 = 0x1c0e07;

const ADJACENT_CELLS: [u128; NUM_OF_CELLS] = {
    let mut tbl = [0_u128; NUM_OF_CELLS];

    let mut pos: usize = 0;
    while pos < NUM_OF_CELLS {
        let row_shift = (pos / 9) * 9;
        let col_shift = pos % 9;
        let box_shift = (pos / 27) * 27 + (col_shift / 3) * 3;
        tbl[pos] |= ROW_MASK << row_shift;
        tbl[pos] |= COL_MASK << col_shift;
        tbl[pos] |= BOX_MASK << box_shift;
        pos += 1;
    }
    tbl
};

fn solve() -> String {
    compute(FILE_DATA).to_string()
}

fn compute(data: &str) -> usize {
    let puzzles = parse_data(data);
    let grids = check_data(&puzzles);

    let mut ans: usize = 0;
    for grid in grids {
        if let Some(sol) = sudoku_solover(&grid) {
            let mut acc: usize = 0;
            for i in 0..3 {
                let bit: u128 = 1 << i;
                for (pos, elm) in sol.iter().enumerate().skip(1) {
                    if *elm & bit != 0 {
                        acc = acc * 10 + pos;
                        break;
                    }
                }
            }
            ans += acc;
        } else {
            // Every puzzle in this problem has each one unique answer.
            unreachable!();
        }
    }
    ans
}

fn parse_data(data: &str) -> Vec<Vec<usize>> {
    let re_data = Regex::new(r"^[0-9.]").unwrap();
    let re_sep = Regex::new(r"^-").unwrap();

    let mut ret: Vec<Vec<usize>> = Vec::new();
    let mut acc: Vec<String> = Vec::new();

    for line in data.lines() {
        if re_data.is_match(line) {
            acc.push(line.to_string());
        } else if re_sep.is_match(line) {
            continue;
        } else if !acc.is_empty() {
            ret.push(process_data(&acc));
            acc.clear();
        }
    }
    if !acc.is_empty() {
        ret.push(process_data(&acc));
    }
    ret
}

fn process_data(data: &[String]) -> Vec<usize> {
    let re_rplc1 = Regex::new(r"[^0-9.]").unwrap();
    let re_rplc2 = Regex::new(r"\.").unwrap();

    let mut tmp = data
        .iter()
        .fold(String::new(), |acc, x| format!("{}{}", acc, x));
    tmp = re_rplc1.replace_all(&tmp, "").to_string();
    tmp = re_rplc2.replace_all(&tmp, "0").to_string();
    tmp.chars().map(|c| c as usize - '0' as usize).collect()
}

fn check_data(pazzles: &[Vec<usize>]) -> Vec<Vec<u128>> {
    let mut ret: Vec<Vec<u128>> = Vec::new();

    'pazzle: for (seq, data) in pazzles.iter().enumerate() {
        if data.len() != NUM_OF_CELLS {
            println!(
                "[Warning] size of input data is mismatch [Grid: {}] (Ignored)\n{:?}",
                seq + 1,
                data
            );
            continue;
        }

        let mut grid: Vec<u128> = vec![0; 10];
        for (pos, &num) in data.iter().enumerate() {
            if num != 0 && grid[num] & ADJACENT_CELLS[pos] != 0 {
                println!(
                    "[Warning] there is a same number in the adjacent cells [Grid: {}] (Ignored)\n{:?}",
                    seq + 1,
                    data
                );
                continue 'pazzle;
            }
            grid[num] |= 1 << pos;
        }
        ret.push(grid);
    }
    ret
}

fn sudoku_solover(grid: &[u128]) -> Option<Vec<u128>> {
    let mut work_grid = grid.to_vec();

    // Start searching from cell #0.
    if _sudoku_solver(&mut work_grid, 0, 0b01) {
        Some(work_grid)
    } else {
        None
    }
}

fn _sudoku_solver(grid: &mut [u128], mut pos: usize, mut pos_bit: u128) -> bool {
    // Search for undetermined cell.
    // If we don't find it, we've found a solution.
    loop {
        // Found a solution
        if pos >= NUM_OF_CELLS {
            return true;
        }

        // Undetermined cell found
        if grid[0] & pos_bit != 0 {
            break;
        }
        pos_bit <<= 1;
        pos += 1;
    }

    // Try numbers 1 through 9.
    for n in 1..=9_usize {
        // If there is a same number in adjacent cells, try next number.
        if grid[n] & ADJACENT_CELLS[pos] != 0 {
            continue;
        }

        // Let number 'n' as temporary on the cell, and try next cell.
        grid[n] |= pos_bit;
        if _sudoku_solver(grid, pos + 1, pos_bit << 1) {
            return true;
        }
        // If the number isn't suitable, undo the status and try next number.
        grid[n] &= !pos_bit;
    }

    // Back to the previous cell if all numbers are not suitable.
    false
}

#[cfg(test)]
mod tests {
    use super::compute;

    #[test]
    fn p0096() {
        assert_eq!(compute(super::FILE_DATA), 24702);
    }
}

File: ocaml/src/p0096.ml
(* Project Euler: Problem 96 *)

(*
 * backtracking version (using Algorithm X)
 *
 * Algorithm X (using Dancing Links)
 *   https://en.wikipedia.org/wiki/Algorithm_X
 *   https://en.wikipedia.org/wiki/Dancing_Links
 *
 * library: algo_x (https://github.com/kwj/algo_x)
 *
 * [Grid]
 *     C0 C1 C2 C3 C4 C5 C6 C7 C8
 *    +--------+--------+--------+    R: Row
 *  R0|        |        |        |    C: Column
 *  R1|   B0   |   B1   |   B2   |    B: Box
 *  R2|        |        |        |
 *    +--------+--------+--------+    Cell(x,y) = RxCy
 *  R3|        |        |        |      position = 9x + y
 *  R4|   B3   |   B4   |   B5   |
 *  R5|        |        |        |
 *    +--------+--------+--------+
 *  R6|        |        |        |
 *  R7|   B6   |   B7   |   B8   |
 *  R8|        |        |        |
 *    +--------+--------+--------+
 *
 * [Matrix]
 *        R0C0 R0C1 .. R8C8 | R0#1 R0#2 .. R8#8 R8#9 | C0#1 C0#2 .. C8#8 C8#9 | B0#1 B0#2 .. B8#8 B8#9
 * ----------------------------------------------------------------------------------------------------
 * R0C0#1   1    0  ..   0      1    0  ..   0    0      1    0  ..   0    0      1    0  ..   0    0
 * R0C0#2   1    0  ..   0      0    1  ..   0    0      0    1  ..   0    0      0    1  ..   0    0
 *  ...
 * R0C0#9   1    0  ..   0      0    0  ..   0    0      0    0  ..   0    0      0    0  ..   0    0
 * R0C1#1   0    1  ..   0      1    0  ..   0    0      0    0  ..   0    0      1    0  ..   0    0
 *  ...
 * R0C8#9   0    0  ..   0      0    0  ..   0    0      0    0  ..   0    1      0    0  ..   0    0
 * R1C0#1   0    0  ..   0      0    0  ..   0    0      1    0  ..   0    0      0    0  ..   0    0
 *  ...
 * R8C8#8   0    0  ..   1      0    0  ..   1    0      0    0  ..   1    0      0    0  ..   1    0
 * R8C8#9   0    0  ..   1      0    0  ..   0    1      0    0  ..   0    1      0    0  ..   0    1
 *
 *   Row: RxCy#N -> Cell(x,y) = N
 *   Col: RxCy -> some number is in Cell(x,y)       [cell constraint] (81 columns)
 *        Rx#N -> number 'N' is in the row Rx       [row constraint] (81 columns)
 *        Cy#N -> number 'N' is in the column Cy    [column constraint] (81 columns)
 *        Bz#N -> number 'N' is in the box Bz       [box constraint] (81 columns)
 *
 * We will need the following files to run this program.
 *   - https://projecteuler.net/project/resources/p096_sudoku.txt
 *)

open Core

let parse_data data =
  let trim s =
    Str.global_replace (Str.regexp "[^0-9.]") "" s
    |> Str.global_replace (Str.regexp "\\.") "0"
  in
  let rec loop lines acc result =
    match lines with
    | [] ->
      List.rev (List.map ~f:trim (List.fold ~f:( ^ ) ~init:"" (List.rev acc) :: result))
    | x :: xs ->
      if Str.string_match (Str.regexp "^[0-9.]") x 0
      then loop xs (x :: acc) result
      else if Str.string_match (Str.regexp "^-") x 0 || List.length acc = 0
      then loop xs acc result
      else loop xs [] (List.fold ~f:( ^ ) ~init:"" (List.rev acc) :: result)
  in
  loop data [] []
;;

let make_dlx q =
  (* number of columns = 324 (81 + 81 + 81 + 81) *)
  let d = Euler.Algo_x.dlx_init 324 in
  let add_row pos n =
    let r, c, b = (pos / 9, pos mod 9, (pos / 27 * 3) + (pos mod 9 / 3)) in
    let add_row' num =
      let tag = Printf.sprintf "R%dC%d#%d" r c num in
      Euler.Algo_x.dlx_add_row
        ~name:tag
        d
        [ pos + 1
        ; 81 + 1 + (r * 9) + (num - 1)
        ; (81 * 2) + 1 + (c * 9) + (num - 1)
        ; (81 * 3) + 1 + (b * 9) + (num - 1)
        ]
    in
    if n <> 0
    then add_row' n
    else
      for i = 1 to 9 do
        add_row' i
      done
  in
  List.iteri
    ~f:(fun i n -> add_row i n)
    (List.map ~f:Int.of_string (Str.split (Str.regexp "") q));
  d
;;

let compute str_lst =
  let pick_num lst =
    let rec loop n l acc =
      if n = 0
      then acc
      else (
        match l with
        | [] -> assert false
        | x :: xs ->
          loop
            (pred n)
            xs
            ((acc * 10)
             + Int.of_string (List.hd_exn (List.rev (Str.split (Str.regexp "") x)))))
    in
    loop 3 lst 0
  in
  let rec loop q i acc =
    match q with
    | [] -> acc
    | x :: xs ->
      let d = make_dlx x in
      (match Euler.Algo_x.dlx_solve d with
       | None -> loop xs (succ i) acc
       | Some l ->
         let ans = pick_num (List.sort ~compare:String.compare (List.hd_exn l)) in
         loop xs (succ i) (acc + ans))
  in
  loop (parse_data str_lst) 1 0
;;

let solve fname = compute (Euler.Task.read_file fname) |> Int.to_string

(* Test *)

let%test_unit "p096_roman.txt" =
  [%test_eq: int] (compute (Euler.Task.read_file "./assets/p096_sudoku.txt")) 24702
;;
